### PR TITLE
Set limit of DHDD to 256 KiB

### LINKED ISSUE #1065: Entities with highspeed can be used to crash the server
Can be used to lag/crash a server.

### PR DESCRIPTION
**Pull request by suunrider**
Fixes #1065.

EDIT by @TomyLobo: 
I recommend using this link to review it:
https://github.com/wiremod/wire/pull/1075/files?w=1

### CODE REVIEW

**Comment by AbigailBuccaneer** [comment]
This limit should be documented somewhere, presumably in the tool description.

**Comment by AbigailBuccaneer** [comment]
(Also, see https://github.com/blog/1506-closing-issues-via-pull-requests for how to format your PRs to automatically close an issue when merged.)

**Comment by Divran** [comment]
what happens when you use address 0.001 or 33.3123551

**Comment by suunrider** [comment]
nothing

**Comment by Divran** [comment]
I've checked the code and I can't find any math.floor anywhere, so it seems
to me that you can bypass the 1mb limit simply by using decimals

**Comment by suunrider** [comment]
I never saw flooring the address in ReadCell. Noticed that you can use floating point values as address on almost every entity with ReadCell/WriteCell. So RAM (1kb) has pretty much storage..

File: lua/entities/gmod_wire_dhdd.lua [hunk 1]
```diff
@@ -23,17 +23,25 @@ end
...
+		local data = self.Memory[Address or 0] or 0
+		return isnumber(data) and data or 0
+	end
 end
 
 -- Write cell
 function ENT:WriteCell( Address, value )
-	if self.AllowWrite then
-		self.Memory[Address] = value
+	if Address < 0 or Address >= 1048576 then -- 1 MiB
```
Comment (AbigailBuccaneer) [inline lua/entities/gmod_wire_dhdd.lua]: This uses `>=` but `ReadCell` uses `>`. Any reason why?

Comment (suunrider) [inline lua/entities/gmod_wire_dhdd.lua]: No :S

**Comment by oldmud0** [comment]
That should be fixed as well, then.

**Comment by TomyLobo** [comment]
You can turn those ifs into short-cut ifs to further minimize the diff.
No need to have an else branch if the if branch returns immediately.

**Comment by suunrider** [comment]
I prefer this way

**Comment by oldmud0** [comment]
Also, this limit should not be hardcoded. It should be set as a config option.

**Comment by AbigailBuccaneer** [comment]
Interestingly, the [DHDD page on the Wiremod Wiki](http://wiki.wiremod.com/wiki/DHDD) says it stores 256 kilobytes.

**Comment by suunrider** [comment]
It saves the first 256 kilobytes in the dupe, rest will be just ignored.

**Comment by AbigailBuccaneer** [comment]
Then given its purpose is a *dupable hard disk drive*, it shouldn't let you add and work with data that can't be duped.

**Comment by Divran** [comment]
Should we merge this?

**Comment by TomyLobo** [comment]
> Should we merge this?

There's still one open issue:

> > You can turn those ifs into short-cut ifs to further minimize the diff.
> > No need to have an else branch if the if branch returns immediately.
>
> I prefer this way

Should a contributor's preference take precedence over readability?

**Comment by Divran** [comment]
>Should a contributor's preference take precedence over readability?

nah

**Comment by suunrider** [comment]
Done

**Comment by TomyLobo** [comment]
One more thing i just noticed: you have an off-by-one error there.
The number of permissible addresses is actually 256*1024+1 now, meaning it's more than 256 kb.
You need to change those ">" operators defining the upper limit into ">=".

**Comment by suunrider** [comment]
Fixed

**Comment by TomyLobo** [comment]
Looks good now

**Comment by bigdogmat** [comment]
Might want to change name of pr of future reference

**Pull request merged**